crates/sui-core/src/execution_cache/writeback_cache.rs

```diff
@@ -0,0 +1,1158 @@
+use sui_types::sui_system_state::{get_sui_system_state, SuiSystemState};
+use sui_types::transaction::VerifiedTransaction;
+use tracing::instrument;
+
+use super::ExecutionCacheAPI;
+use super::{
+    cached_version_map::CachedVersionMap, implement_passthrough_traits, CheckpointCache,
+    ExecutionCacheCommit, ExecutionCacheMetrics, ExecutionCacheRead, ExecutionCacheReconfigAPI,
+    ExecutionCacheWrite, NotifyReadWrapper, StateSyncAPI,
+};
+
+#[cfg(test)]
+#[path = "unit_tests/memory_cache_tests.rs"]
+mod authority_tests;
+
+#[derive(Clone, PartialEq, Eq)]
+enum ObjectEntry {
+    Object(Object),
+    Deleted,
+    Wrapped,
+}
+
+impl std::fmt::Debug for ObjectEntry {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            ObjectEntry::Object(o) => {
+                write!(f, "ObjectEntry::Object({:?})", o.compute_object_reference())
+            }
+            ObjectEntry::Deleted => write!(f, "ObjectEntry::Deleted"),
+            ObjectEntry::Wrapped => write!(f, "ObjectEntry::Wrapped"),
+        }
+    }
+}
+
+impl From<Object> for ObjectEntry {
+    fn from(object: Object) -> Self {
+        ObjectEntry::Object(object)
+    }
+}
+
+type MarkerKey = (EpochId, ObjectID);
+
+enum CacheResult<T> {
+    /// Entry is in the cache
+    Hit(T),
+    /// Entry is not in the cache and is known to not exist
+    NegativeHit,
+    /// Entry is not in the cache and may or may not exist in the store
+    Miss,
+}
+
+/// UncommitedData stores execution outputs that are not yet written to the db. Entries in this
+/// struct can only be purged after they are committed.
+struct UncommittedData {
+    /// The object dirty set. All writes go into this table first. After we flush the data to the
+    /// db, the data is removed from this table and inserted into the object_cache.
+    ///
+    /// This table may contain both live and dead objects, since we flush both live and dead
+    /// objects to the db in order to support past object queries on fullnodes.
+    /// When we move data into the object_cache we only retain the live objects.
+    ///
+    /// Further, we only remove objects in FIFO order, which ensures that the the cached
+    /// sequence of objects has no gaps. In other words, if we have versions 4, 8, 13 of
+    /// an object, we can deduce that version 9 does not exist. This also makes child object
+    /// reads efficient. `object_cache` cannot contain a more recent version of an object than
+    /// `objects`, and neither can have any gaps. Therefore if there is any object <= the version
+    /// bound for a child read in objects, it is the correct object to return.
+    objects: DashMap<ObjectID, CachedVersionMap<ObjectEntry>>,
+
+    // Markers for received objects and deleted shared objects. This contains all of the dirty
+    // marker state, which is committed to the db at the same time as other transaction data.
```

Comment:
I'm sure this will become apparent to me as I continue to read, but just want to double-check that this commit is either atomic or that these markers will be committed before the object updates

Reply:
it is not atomic! but the markers are written before the object updates.

Reply:
Ah! thank you for the correction! For some reason when I saw the call to `write_transaction_outputs` I got the atomicity from there, but you are indeed correct that the "shuffling" between caches is not atomic. Markers gettin written before object updates is the main thing :)